README
Delorean (DELete Often, the system REtAiN) - a time travel file system
----------------------------------------------------------------------

Rafael Ignacio Zurita, 2008-2010


Introducción
============

Delorean es un sistema de archivos que mantiene las diferentes versiones de los
archivos a través del tiempo.

Además, y para poder cumplir con ciertas metas pre-fijadas, Delorean implementa
la útil característica de poder volver atrás en el tiempo a nivel de procesos 
del sistema operativo.
Esto significa que diferentes procesos pueden compartir el mismo sistema de 
archivos en diferentes puntos en el tiempo.


Compilación
===========

- Antes de compilar se debe tener instalado las librerías de fuse y sus headers.
  Delorean utiliza la librería de FUSE, los header de FUSE y fuse-utils.
  En Debian los paquetes a instalar son: fuse-utils, libfuse-dev y
  libfuse2.

- ejecute 'tar xvzf delorean.tar.gz' para desempaquetar los archivos fuentes.
- cd delorean/

- ejecutar 'make clean ; make all ' para compilar


Testeando delorean
==================

Para montar y utilizar un sistema de archivos delorean se debe cargar el módulo
del kernel fuse. Ejecutar 'modprobe fuse' si no se encuentra ya cargado.
También es necesario crear el directorio para las versiones y el punto de 
montura del sistema de archivos. Una vez preparado los directorios utilice
mount.delorean para montar el sistema de archivos.

Por lo que los comandos a ejecutar serían:
	- modprobe fuse
	- mkdir -p MOUNT_PATH/.versions+place ; mkdir MOUNT_POINT
	- ./mount.delorean MOUNT_POINT -o default_permissions -f
	- cd MOUNT_POINT/
	- (utilizar..)

Lea el siguiente ejemplo concreto si aún quedan dudas ..:
# mkdir -p /mnt/fuse/real/.versions+place
# mkdir /mnt/fuse/mount_point
# ./mount.delorean /mnt/fuse/mount_point/ -o default_permissions -f

El shell utilizado para montarl el sistema de archivos queda en modo debug,
utilice otro(s) shell para utilizar el filesystem.

# cd /mnt/fuse/mount_point/
(utilizar el FS como cualquier otro FS).

Para "desmontar el sistema de archivos delorean utilice umount como siempre.
Para el ejemplo anterior sería 'umount /mnt/fuse/mount_point'


FAQ
===
1. ¿Cómo eliminar las versiones guardadas?
2. ¿Cómo utilizar delorean en un momento pasado?
3. ¿Cómo traer un archivo que ya no existe (o varios) al presente?
4. ¿Otro ejemplo de utilización?
5. ¿Cómo realizar un backup del FS en un punto en el tiempo pasado? (snapshot)
6. ¿Cuál es la Granularidad?
7. ¿Porqué algunos programas fallan en el pasado?


1. ¿Cómo eliminar las versiones guardadas?

Sencillo. Desmontar y eliminar las versiones.
Ejemplo:
# umount /mnt/fuse/mount_point
# rm -rf /mnt/fuse/real/.versions+place/*


2. ¿Cómo utilizar delorean en un momento pasado?

cd MOUNT_POINT
travel año-mes-dia hora:min:seg commando

Ejemplo:
# cd /mnt/fuse/mount_point
# travel 2007-09-07 12:34:20 /bin/bash
.. y un nuevo bash estará viajando ;-)

(El comando anterior ejecuta un nuevo shell bash en el pasado. A la fecha 
07 de Septiembre de 2007 a las 12 del mediodía, 34 minutos y 20 segundos).

El shell y todos los descendientes de ese shell verán el FS al momento pasado
mencionado.


3. ¿Cómo traer un archivo que ya no existe (o varios) al presente?
Un simple ejemplo:

# date; ls
Tue Sep  4 22:37:59 ART 2007
sex_pistols.ogg	fat_boy_slim.ogg
# sleep 10; date; rm *.ogg
Tue Sep  4 22:38:20 ART 2007
# ls
#
# travel 2007-09-04 22:38:05 /bin/bash
# ls
sex_pistols.ogg	fat_boy_slim.ogg
# cp *.ogg /tmp
# exit
# cd /tmp
# ls
sex_pistols.ogg	fat_boy_slim.ogg

En el ejemplo anterior se copiaron dos archivos ogg (que ya no existían) al 
directorio /tmp (el cual está en el presente).


4. ¿Otro ejemplo de utilización?

# date; ls
Tue Sep  4 22:37:59 ART 2007
sex_pistols.mp3	fat_boy_slim.mp3
# sleep 10; date; rm *.mp3
Tue Sep  4 22:38:20 ART 2007
# ls
#
# travel 2007-09-04 22:38:05 /bin/bash
# ls
sex_pistols.mp3	fat_boy_slim.mp3
# mplayer *.mp3

:)

Advertencia:	a veces un viaje al pasado confunde :).. Desde que algunos
		archivos actuales tambien fueron los mismos en el pasado, se 
		mezclan con archivos del pasado que ya no existen. Y algunas
		veces existe la sensación de que el FS devuelve información
		incorrecta. En general, cuando uno chequea quien está, quien se 
		fue, quien está y estuvo .. todo se aclara.


5. ¿Cómo realizar un backup del FS en un punto en el tiempo pasado? (snapshot)
Muy pavo a esta altura... Desde los ejemplos anteriores:

# date; ls
Tue Sep  4 22:37:59 ART 2007
sex_pistols.mp3	fat_boy_slim.mp3
# sleep 10; date; rm *.mp3
Tue Sep  4 22:38:20 ART 2007
# ls
#
# travel 2007-09-04 22:38:05 /bin/bash
# ls
sex_pistols.mp3	fat_boy_slim.mp3
# tar cvzf /tmp/backup.tar.gz .


6. ¿Cuál es la Granularidad?

El sistema de archivos guarda como máximo una versión por segundo de cada
archivo que se modifica.

Cada nueva operación open guarda una copia del archivo (si es que no existe
ya algún proceso que tenga un descriptor a ese archivo abierto). Cuando
se realiza el último close (un release en donde no queden procesos con descriptores
a ese archivo abierto) se comparan las versiones. Si las versiones difieren, el FS
mantiene la copia. Si no la descarta.

Por lo tanto el sistema de archivos guarda una única versión entre el primer
open a un archivo y su ultimo close (release). Esto significa que si luego del
primer open y antes del ultimo close (release) existen varias operaciones
open/read/write/close el sistema de archivos no realizará nuevos respaldos de
versiones.


7. ¿Porqué algunos programas fallan en el pasado?

El sistema de archivos es de solo lectura cuando un proceso lo utiliza en un
punto de tiempo pasado. Por lo que programas que necesiten generar temporales
o escribir directamente deberían reportar al usuario que no pueden escribir.

Si los procesos utilizan el número de inodo de los archivos para realizar su
tarea pueden "confundirse" :P .. Los números de inodos no son mantenidos en las
versiones respaldadas, por lo que extrañan situaciones pueden suceder con tales
programas :)

Un move de un archivo en el sistema de archivos actual va a devolver como 
resultado dos archivos en el pasado. ¿Por qué sucede esto?... Bien, esta 
situación se da porque el archivo movido mantiene su fecha de modificación 
(fecha de modificación del contenido), y luego, en un punto del tiempo
pasado el FS va a detectar esas dos versiones como si hubieran "vivido" en el
mismo momento (la versión original antes del move va a existir para el 
sistema de archivos hasta el momento en que fue movido, y para la versión
destino del move delorean "interpreta" que nació en el mismo momento que 
la versión original y vive o vivió en un futuro posterior al move). Suena
raro. Lo mejor es realizar una prueba sencilla de mover un archivo a otro
directorio. Y viajar con un shell a un punto anterior al move. delorean
devolverá ambas versiones... No me gusta esa forma de trabajar.. Habrá que
mejorarla.

Si se mueve un directorio en el sistema de archivos actual no se guardan
versiones. Y en un punto del tiempo anterior al move el sistema de archivos
no va a detectar esas versiones.. :( Esto es así porque el FS chequea si el contenido
de los archivos cambia, pero no lo relacionado con el contenido de los directorios
o de los inodos. Cuando un proceso utiliza el sistema de archivos en un punto 
de tiempo pasado delorean intenta deducir que directorios existían por conocer
donde residían los archivos de los cuales si tiene conocimiento en cuanto a
tiempo de vida. 

Leer el código para comprender todas estas respuestas ;-)





Changelog
Lista de cambios informal
=========================

Version testing actual 0.4

Version de desarrollo 0.1.28

0.4
---

Se corrige el chequeo de ddir_flag en fc_getdir() en flux_capacitor.c; el
cual omitia el if original and generaba una salida incorrecta 
(utilizando la opcion else). Chequear un diff para conocer mas detalles.

0.3 
---

Desde la primera version testing 0.1 (la cual fue lo suficientemente funcional
para pruebas) se hicieron los siguiente cambios :

* Se corrige la funcion get_fn() para que reserve (malloc) la cantidad
  apropiada de memoria a utilizar.

* Se corrige la funcion get_fn() para que si f == / entonces simplemente
  salga de la funcion y devuelva /

* Se corrige el bug en travel que utilizaba incorrectamente la estructura
  tm de time.h. En travel no se inicializaba completamente tal estructura,
  por lo que en algunos sistemas Linux el timestamp utilizado con travel
  era modificado por el daylight saving time del sistema.
  Reportado y testeado por Claudio Vaucheret (Universidad Nacional del Comahue).


test_delorean.sh
#!/bin/bash

export DISPLAY=:0
xhost +
rm -rf real
rm -rf mp
mkdir -p real/.versions+place
mkdir mp
mount.dlorean real mp -o default_permissions -f -d


list.c
/*
 * list.c - very stupid list. It is part of Dlorean project.
 * Rafael Ignacio Zurita, 2007
 */

#include <stdio.h>
#include <stdlib.h>
#include <sys/queue.h>

#include "list.h"


/*
 * add()
 * add an element to the list
 * the element has 2 strings and an integer:
 * 1- the file name being versioned
 * 2- the file name used as saved version
 * 3- the int n times that the file was opened
*/
int add (const char *f1, char *f2){
    size_t f1len=(strlen(f1)+1);
    size_t f2len=(strlen(f2)+1);
    void *res;
    struct entry *n1;

    n1 = malloc(sizeof(struct entry));      /* Insert at the head. */
    n1->fname=malloc(f1len);
    n1->bfname=malloc(f2len);
    n1->n = 1;
    res=memcpy(n1->fname,f1,f1len);
    res=memcpy(n1->bfname,f2,f2len);

    
    LIST_INSERT_HEAD(&head, n1, entries);
    printf("dato insertado=_%s_:_%s_\n",n1->fname,n1->bfname);

    return 1;
}


/*
 * exist():
 *  if f exists and n1->n==1 and c==-1 then return 1
 *  if f exists and n1->n>1 and c==-1 then n1-> -c 
 *  if f exists and c==1 then n1->n + c 
 *  if not, then return 0
 *  c should be 1 or -1
 *  it is because this function is usefull when exist un "open" or a "close"
 *  operation
*/

int exist (const char *f){
    struct entry *n1;

    for (n1 = head.lh_first; ((n1 != NULL)&&(strcmp(n1->fname,f))); n1 = n1->entries.le_next);
    if ( (n1!=NULL)&&(!strcmp(n1->fname,f)) )
	    return n1->n;

    return 0;
};



int mod (const char *f,int c){
    struct entry *n1;

    for (n1 = head.lh_first; n1 != NULL; n1 = n1->entries.le_next)
	     if (!strcmp(n1->fname,f)) {
		     n1->n = n1->n + c;
		     return 1;
	     };
    printf("[CASO EXTRANIO]\n");
    return 0;
};


/*
 * query()
 * query in the list: 
 * if file (*f) exists return it
 * if not, return 0
*/
char *query (const char *f, int *n){
    struct entry *n1;
    char *s;
printf("[file to compare:%s]\n",f);
    for (n1 = head.lh_first; n1 != NULL; n1 = n1->entries.le_next){
printf("[file to compare2:%s]\n",n1->fname);
	     if (!strcmp(n1->fname,f)) {
		     printf("[elemento1:%s]\n",n1->fname);
		     printf("[elemento2:%s]\n",n1->bfname);
		     s = malloc(strlen(n1->bfname)+1);
		     memcpy(s,n1->bfname,strlen(n1->bfname)+1);
		     *n=n1->n;
		     return s;
	     };
    };
    *n=0;
    return 0;
};


int del (const char *f){
    struct entry *n1;

    for (n1 = head.lh_first; ((n1 != NULL)&&(strcmp(n1->fname,f))); n1 = n1->entries.le_next);
    if (!strcmp(n1->fname,f)) {
	LIST_REMOVE(n1, entries);
	free(n1->fname);
	free(n1->bfname);
	free(n1);
	return 1;
    };
	
    return 0;
};



list.h

/*
 * list.c - very stupid list. It is part of Dlorean project.
 * Rafael Ignacio Zurita, 2007
 */

#ifndef LIST_H
#define LIST_H


#include <stdio.h>
#include <stdlib.h>
#include <sys/queue.h>


LIST_HEAD(listhead, entry) head;
struct listhead *headp;	             /* List head. */
struct entry {
     char *fname;           /* the file name being versioned */
     char *bfname;          /* the file name used as version saved */
     int   n;               /* n times opening */
     LIST_ENTRY(entry) entries;      /* List: open files. */
}; 


/*
 * add()
 * add an element to the list
 * the element has 2 strings and an integer:
 * 1- the file name being versioned
 * 2- the file name used as saved version
 * 3- the int n times that the file was opened
*/
int add (const char *f1, char *f2);

/*
 * exist():
 *  if f exists and n1->n==1 and c==-1 then return 1
 *  if f exists and n1->n>1 and c==-1 then n1-> -c 
 *  if f exists and c==1 then n1->n + c 
 *  if not, then return 0
 *  c should be 1 or -1
 *  it is because this function is usefull when exist un "open" or a "close"
 *  operation
 */

int exist (const char *f);

int mod (const char *f,int c);


/*
 * query()
 * query in the list: 
 * if file (*f) exists return it
 * if not, return 0
*/
char *query (const char *f, int *n);

int del (const char *f);

#endif


travel.c
/*
 * travel.c - registers/unregisters pids in dlorean traveling pids table,
 *            and executes a new program.
 * It is part of Dlorean project
 * Rafael Ignacio Zurita, 2007
 */

#include <sys/wait.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <signal.h>
#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include <time.h>
#include <fcntl.h>

#include "flux_capacitor.h"

int
main(int argc, char *argv[])
{
	pid_t cpid, w;
	int status,i,err;
	time_t ts,tshelp;
	struct tm tmstruct;
	char *chartime=malloc(20);
	char *fn=malloc(16);  /* fn: temp file name. E.g. /tmp/1235678901 */
	char *fnreg=malloc(60); /* fnreg: file name for pid registration */
	char *fnreg2=malloc(60); /* fnreg2: file name for unregistration pid */
	struct stat buf;

	if (!(argc == 4)){
		printf("\nUse:\n\t# travel yyyy-mm-dd hh:mm:ss command\n\n");
		exit(1);
	};

	char *command=malloc(strlen(argv[3])+1);;
	strcat(command,argv[3]);
	printf("Ejecutando %s ...\n",command);

	memcpy(chartime,argv[1],strlen(argv[1]));
	memcpy(chartime+10,argv[2],strlen(argv[2])+1);
	strptime(chartime, "%Y-%m-%d%H:%M:%S", &tmstruct);

	tmstruct.tm_isdst = -1;
	ts = mktime(&tmstruct);           

	tshelp = time(NULL);
	sprintf(fn,"/tmp/%u",tshelp);

	cpid = fork();

	/*
	 * llamar al sistema de archivos con un stat de .traveling(segundos
	 * desde 1970)(pid del proceso a viajar)
	 */
	sprintf(fnreg,"%s%u%i",reg_pid,ts,cpid);
	sprintf(fnreg2,"%s%u%i",unreg_pid,ts,cpid);

	if (cpid == -1) { perror("fork"); exit(EXIT_FAILURE); }

	if (cpid > 0) {		       /* Code executed by parent */
	        /* 
		 * Intento registrar en el sistema de archivos
		 * al proceso hijo que va a comenzar a viajar 
		 */
		memset(&buf,0,sizeof(buf));
		err = stat(fnreg, &buf);
		if (err==-1) {
			printf("Error al registrar proceso viajante. Abortando ..\n");
			kill(cpid,SIGKILL);
			exit(err);
		}

		err = mknod(fn, S_IFREG, 0);  /* dev=0 ignored */
		if (err==-1) {
			printf("Error al crear archivo de coordinacion. Abortando ..\n");
			kill(cpid,SIGKILL);
			exit(err);
		}
		/* Espero a que finalice la ejecucion de mi proceso hijo */
		do {
			w = waitpid(cpid, &status, WUNTRACED | WCONTINUED);
		} while (!WIFEXITED(status) && !WIFSIGNALED(status)); 

		/* aca debo desregistrar en el sistema de archivos al proceso hijo */
		memset(&buf,0,sizeof(buf));
		err = stat(fnreg2, &buf);
		if (err==-1) {
			printf("Error al desregistrar proceso viajante. Abortando ..\n");
			kill(cpid,SIGKILL);
			exit(err);
		} 

		printf("Ya ha terminado la ejecucion de %s\n",command);
		err = unlink(fn);
		exit(EXIT_SUCCESS);
	}
	else { 	       /* Code executed by child */
	       
		/* espero a que mi padre registre mi PID en el sistema de archivos */
		do {
			memset(&buf,0,sizeof(buf));
			err = stat(fn, &buf);
		} while (err < 0);
		execve(command,NULL,NULL);
	}
}


flux_capacitor.h
#ifndef FLUX_CAPACITOR_H
#define FLUX_CAPACITOR_H

/*
 * flux_capacitor.c - saves and retrieves file versions in the file system,
 *                    registers/unregisters proccess pids in the traveling pids
 *                    table and other useful functions.
 *
 * It is part of Dlorean project
 * Rafael Ignacio Zurita, 2006
 */

#define _FILE_OFFSET_BITS 64
#define FUSE_USE_VERSION 26

#include <stdio.h>
#include <string.h>
#include <time.h>
#include <libgen.h>

#include "list.h"

#include <stdlib.h>
#include <errno.h>
#include <fcntl.h>

#include <dirent.h>
#include <sys/stat.h>
#include <sys/statfs.h>
#include <sys/types.h>
#include <unistd.h>
#include <utime.h>
#include <fuse.h>


/* Modify real_path if you want to use another directory as mount path */
//static const char *real_path = "/mnt/fuse/real";
char *real_path;

//static const char *mount_point_path = "/mnt/fuse/mount_point";
char *mount_point_path;
static const char *versions_place = ".versions+place";

static const char *reg_pid = ".register";
static const char *unreg_pid = ".unregister";

char *save(const char *file);

struct traveler {
	int pid;
	int time;
	struct traveler *next;
} *traveling;


/*
 * useful strcatloc given by wa
 */
char *strcatloc(const char *a,const char *b);

/*
 * reemplaza el principio de la cadena old por new
 */
char *strreplace(const char *a,const char *old, const char *new);

/*
 * check_and_save - guarda una copia del archivo file
 * (si es que aun no ha sido respaldada)
 * TODO: reeemplazar return 1 por return 0/-1 :P
 */
int check_and_save(const char *file);

char *save(const char *file);


/*
 * cmp_and_confirm - compara la copia guardada con el archivo actual. Si son
 * diferentes mantiene la copia. Si son iguales la elimina
 * (si aún existen procesos que mantienen algún descriptor para file entonces
 * se decrementa el contador en la lista de archivos abiertos)
 * TODO: reemplazar return 1 por retunr 0/-1 
 */
int cmp_and_confirm(const char *file);


/* useful functions to register - unregister pids */
int traveling_pid (int pid);

int register_pid (int pid, int time);

int unregister_pid (int pid);

/*
 * Por ahora no es necesaria print_pids()
 * void print_pids (void)
 * {
 * 	struct traveler *t;
 * 	if (traveling == NULL)
 * 		return -1;
 * 	else {
 * 		t = traveling;
 * 		while(t->next!=NULL){
 * 			printf("pid=%i....time%i\n",t->pid,t->time);
 * 			t = t->next;
 * 		}
 * 		printf("pid=%i....time%i\n",t->pid,t->time);
 * 		return 0;
 * 	}
 * }
 */

/* END of useful functions for register - unregister pid */


/* 
 * systems calls functions for traveling pids 
 * these are called when travel.c wants to register a pid
 */

int fc_getattr(const char *path);
	
/* END of systems calls functions for traveling pids */


/*
 * in_the_past is a pswalk() - walks across parents until to find a traveling
 * pid.
 * I use some code from pstree.c 
 */

#define PROC_BASE    "/proc"
int in_the_past(int pid);

/*
 * file_at_time checks if f and fc have the same name(without timestamp), and if 
 * f has been living at time t.
 * for example: f=filename, fc=filename.1234567890, t=1234567890 (timestamp is .1234567890)
 * second example: f=filename.1234567890, fc=filename.1234567890, t=1234567890
 */
int file_at_time(char *f, char *fc, int t, int check_ctime);


/*
 * dir_at_time checks if dir d existed at time t
 */
int dir_at_time(char *d, int t, int check_ctime);


/* 
 * get_fn(f,t): Devuelve el nombre del archivo real para el tiempo t (solo para pids viajando)
 * f = file. t = time
 * get_fn devuelve el nombre del archivo que el pid que esta viajando debería ver.
 * Por ej.: si el proceso que viaja al pasado (al pasado "t") desea realizar un open(/dir/file)
 * get_fn("/dir/file", t) deberia devolver el nombre del archivo respaldado que 
 * contiene la información que existió en ese archivo para el tiempo "t".
 * El tiempo t es el tiempo en que el pid está viajando.
 */

char *get_fn(const char *f, int t);


/*
 * check if dir2 lives in the same FS than dir1
 */
int samefs(char *dir1, char *dir2);


/*
 * this du estimates the .versions+place size.
 * and that size should be used in statfs at least.
 */

int du(char *dir);

int not_included (int n, struct dirent **namelist, char *file);

//int fc_getdir(int ts, const char *path, fuse_dirh_t h, fuse_dirfil_t filler)
int fc_getdir(int ts, const char *path, void *h, fuse_fill_dir_t filler);
 
#endif


dlorean.c
/*
 *  dlorean.c : Un versioning filesystem orientado al usuario
 *
 *  Copyright (c) 2009 - Rafael Ignacio Zurita <rizurita@yahoo.com>
 *
 *
 *  Leer el archivo README para conocer como compilar y utilizar
 *
 *  This program can be distributed under the terms of the GNU GPL.
 *  See the file COPYING.
 */ 

#define _FILE_OFFSET_BITS 64
#define FUSE_USE_VERSION 26
#include <fuse.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <fcntl.h>

#include <dirent.h>
#include <sys/stat.h>
#include <sys/statfs.h>
#include <sys/types.h>
#include <unistd.h>
#include <utime.h>

#include "flux_capacitor.h"


static int dlorean_getattr(const char *path, struct stat *stbuf)
{
	char *path_t;
	int res = 0;
	int time;
	
	if (!fc_getattr(path)){
		printf("devuelvo 0!..\n");
		res = lstat("/dev/null",stbuf);
		return 0;
	};
	
	//  **** PARA NO PERMITIR EL INGRESO AL DIRECTORIO de versiones !!**
	if (!strncmp(versions_place,path+1,strlen(versions_place)))
		    return -2;
	
	if (fuse_get_context()->pid < 1) 
		    return -2;

	if ((time=in_the_past(fuse_get_context()->pid)) > 0) 
		path_t=get_fn(path, time);
	else
		path_t = strcatloc(real_path, path);
	
	// el siguiente memset.. puede ser reemplazado por stbuf[..]=0?
	memset(stbuf, 0, sizeof(struct stat));
	
	// utilizo lstat para poder obtener tambien el stat de symbolic links
	res = lstat(path_t, stbuf);
	
	free(path_t);
	
	// if res<0 then hubo un error con stat. Devuelvo -2 (no such file or dir)
	if (res < 0)
		res = -2;
	return res;
}

static int dlorean_statfs(const char *path, struct statvfs *stbuf)
{
	char *path_t;
	int res = 0;
	
	path_t = strcatloc(real_path, path);
	// el siguiente memset.. puede ser reemplazado por stbuf[..]=0?
	memset(stbuf, 0, sizeof(struct statvfs));
	res = statvfs(path_t, stbuf);
	free(path_t);
	
	return res;
	}
	
static int dlorean_readlink(const char *path, char *buf, size_t size)
{
	char *path_t;
	int res = 0;
	
	path_t = strcatloc(real_path, path);
	res = readlink(path_t, buf, size-1);
	buf[res] = 0;
	free(path_t);
	
	// controlo si en buf existe el camino real_path
	if ( (path_t = strstr(buf, real_path)) && (strlen(path_t)==strlen(buf)) ) {
		path_t = strreplace (buf, real_path, mount_point_path);
		memcpy(buf, path_t, strlen(path_t));
		buf[strlen(path_t)] = 0;
	};
	
	// res = a la cantidad de de caracteres ubicados en el buffer
	// o res = -1 si hubo un error.
	// dlorean_readlink debería devolver 0 si todo ha ido bien.
	if (res > -1)
		res = 0;
	return res;
}

static int dlorean_mknod(const char *path, mode_t mode, dev_t dev)
{
	char *path_t;
	int res = 0;
	
	int time = in_the_past(fuse_get_context()->pid);
	if (time > 0) 
		return -1;
	
	path_t = strcatloc(real_path, path);
	res = mknod(path_t, mode, dev);
	free(path_t);
	
	return res;
}

static int dlorean_utime(const char *path, struct utimbuf *buf)
{
	char *path_t;
	int res = 0;
	
	int time = in_the_past(fuse_get_context()->pid);
	if (time > 0) 
		return -1;
	
	path_t = strcatloc(real_path, path);
	res = utime(path_t, buf);
	free(path_t);
	
	return res;
}

static int dlorean_mkdir(const char *path, mode_t mode)
{
	char *path_t;
	int res = 0;
	
	int time = in_the_past(fuse_get_context()->pid);
	if (time > 0) 
		return -1;
	
	path_t = strcatloc(real_path, path);
	res = mkdir(path_t, mode);
	free(path_t);
	
	return res;
}
	
static int dlorean_unlink(const char *path)
{
	char *path_t;
	int res = 0;
	
	int time = in_the_past(fuse_get_context()->pid);
	if (time > 0) 
		return -1;
	
	char *f;
	f = save(path);    /* in flux_capacitor.c */
	free(f);
	
	path_t = strcatloc(real_path, path);
	res = unlink(path_t);
	free(path_t);
	
	return res;
}

static int dlorean_rmdir(const char *path)
{
	char *path_t;
	int res = 0;
	
	int time = in_the_past(fuse_get_context()->pid);
	if (time > 0) 
		return -1;
	
	path_t = strcatloc(real_path, path);
	res = rmdir(path_t);
	free(path_t);
	
	return res;
}
	
static int dlorean_symlink(const char *oldpath, const char *newpath)
{
	char *path_n, *path_o;
	int res = 0;
	
	int time = in_the_past(fuse_get_context()->pid);
	if (time > 0) 
		return -1;
	
	path_n = strcatloc(real_path, newpath);
	
	// controlo si en buf existe el camino real_path
	if ( (path_o = strstr(oldpath, mount_point_path)) && (strlen(path_o)==strlen(oldpath)) ) {
		path_o = strreplace (oldpath, mount_point_path, real_path);
		res = symlink(path_o, path_n);
		free(path_o);
	} else 
		res = symlink(oldpath, path_n);
	
	free(path_n);
	
	return res;
}

static int dlorean_rename(const char *oldpath, const char *newpath)
{
	char *path_o, *path_n;
	int res = 0;
	
	int time = in_the_past(fuse_get_context()->pid);
	if (time > 0) 
		return -1;
	
	char *f;
	f = save(oldpath);    /* in flux_capacitor.c */
	free(f);
	
	path_o = strcatloc(real_path, oldpath);
	path_n = strcatloc(real_path, newpath);
	res = rename(path_o, path_n);
	free(path_n);
	free(path_o);
	
	return res;
}
	
static int dlorean_link(const char *oldpath, const char *newpath)
{
	char *path_o, *path_n;
	int res = 0;

	int time = in_the_past(fuse_get_context()->pid);
	if (time > 0) 
		return -1;
	
	path_o = strcatloc(real_path, oldpath);
	path_n = strcatloc(real_path, newpath);
	res = link(path_o, path_n);
	free(path_n);
	free(path_o);
	
	return res;
}

static int dlorean_chmod(const char *path, mode_t mode)
{
	char *path_t;
	int res = 0;
	
	int time = in_the_past(fuse_get_context()->pid);
	if (time > 0) 
		return -1;
	
	path_t = strcatloc(real_path, path);
	res = chmod(path_t, mode);
	free(path_t);
	
	return res;
}
	
static int dlorean_chown(const char *path, uid_t owner, gid_t group)
{
	char *path_t;
	int res = 0;
	
	int time = in_the_past(fuse_get_context()->pid);
	if (time > 0) 
		return -1;
	
	path_t = strcatloc(real_path, path);
	res = chown(path_t, owner, group);
	free(path_t);
	
	return res;
}

static int dlorean_truncate(const char *path, off_t newsize)
{
	char *path_t;
	int res = 0;
	
	int time = in_the_past(fuse_get_context()->pid);
	if (time > 0) 
		return -1;
	
	char *f;
	f = save(path);    /* in flux_capacitor.c */
	free(f);
	
	path_t = strcatloc(real_path, path);
	res = truncate(path_t, newsize);
	free(path_t);
	
	return res;
}

static int dlorean_opendir(const char *path, struct fuse_file_info *fi)
{
/*
	int size_c, dir_ok;
	char *path_t;
	time_t ts;
	DIR *n;

	(void) fi;
	
	if ((ts=in_the_past(fuse_get_context()->pid)) > 0) {
		size_c = strlen(real_path) + 1 + strlen(versions_place) + strlen(path);
		path_t = malloc(size_c+1);
		printf(path_t,"%s/%s%s",real_path,versions_place,path);
		if ((dir_ok = dir_at_time(path_t, ts, 1)) != 0)
			return -1;
	}
	else 
		path_t = strcatloc(real_path, path);
	n = opendir(path_t);
	free(path_t);
	if (n == NULL)
		return -1;
	else */
		return 0;

}

//static int dlorean_getdir(const char *path, fuse_dirh_t h, fuse_dirfil_t filler)
static int dlorean_readdir(const char *path, void *h, fuse_fill_dir_t filler,
			  off_t offset, struct fuse_file_info *fi)
{
	struct dirent **namelist;
	int n;
	char *path_t;
	time_t ts;

	(void) offset;
	(void) fi;

	
	ts = in_the_past(fuse_get_context()->pid);
	if (ts > 0) 
		return fc_getdir(ts, path, h, filler);
	else {
		path_t = strcatloc(real_path, path);
		n = scandir(path_t, &namelist, 0, alphasort);
		free(path_t);
	
		if (n < 0) {
			printf("1111: no pude hacer un scandir de:%s__\n",path_t);
			perror("scandir"); /* perhaps should be return 1 */
		} else {
			while (n--) {
				/* 
				 * si namelist[n]->d_name no es verions+place entonces
				 * coloco a namelist[n]->d_name en la estructura h
				 */
				if (strncmp(versions_place,namelist[n]->d_name,strlen(versions_place))){
					struct stat st;
					memset(&st, 0, sizeof(st));
					st.st_ino = namelist[n]->d_ino;
					// st.st_mode..
		    			//filler(h, namelist[n]->d_name, namelist[n]->d_type, namelist[n]->d_ino);
		    			filler(h, namelist[n]->d_name, &st, 0);
				};
		    		free(namelist[n]);
			}
		}
	};
	return 0;
}
static int dlorean_releasedir(const char *path, struct fuse_file_info *fi)
{
		return 0;
}

static int dlorean_release(const char *path, struct fuse_file_info *fi)
{
	int n;

	int time = in_the_past(fuse_get_context()->pid);
	if (time < 0) 
		cmp_and_confirm(path);    /* in flux_capacitor.c */
	
	n = close (fi->fh);
	if (n == -1) {
		perror("close");
		return -EACCES;
	} 
	return 0;
}

static int dlorean_open(const char *path, struct fuse_file_info *fi)
{
	int n;
	char *path_t;
	
	samefs(real_path, mount_point_path);
	/* The line below doesn't work :(. Perhaps i should check only "readonly"
	* flag in an open operation
	*/
	//if ((fi->flags == O_WRONLY) || (fi->flags == O_RDWR))
	/*
	* if ((fi->flags > O_RDONLY))
	*      save_a_copy(path);
	*/
	/* When i open some files (if they already exist) the flags
	* are 32768 (with "more" for example) or > 32768.
	* Is read only 0 only?
	*/
	
	int time = in_the_past(fuse_get_context()->pid);
	if (time > 0)
		path_t=get_fn(path, time);
	else {
		check_and_save(path);
		path_t = strcatloc(real_path, path);
	};
	
	n = open (path_t, fi->flags);
	free(path_t);
	if (n == -1) {
		perror("open");
		return -EACCES;
	} 
	fi->fh = n;
	
	return 0;
}

static int dlorean_write(const char *path, char *buf, size_t size, off_t offset,
                      struct fuse_file_info *fi)
{
	off_t lseek_r;
	size_t len;

	int time = in_the_past(fuse_get_context()->pid);
	if (time > 0) 
		return -1;

	lseek_r = lseek (fi->fh, offset, SEEK_SET);
	// aqui tal vez deberia controlar si lseek fue exitoso o no
	len = write (fi->fh, buf, size);
	if (len == -1) {
		perror("write");
		return -1;
	}
	
	return len;
}

static int dlorean_read(const char *path, char *buf, size_t size, off_t offset,
                      struct fuse_file_info *fi)
{
	off_t lseek_r;
	size_t len;
	
	lseek_r = lseek (fi->fh, offset, SEEK_SET);
	// aqui tal vez deberia controlar si lseek fue exitoso o no
	len = read (fi->fh, buf, size);
	if (len == -1) {
		perror("read");
		return -1;
	}
	
	return len;
}

static struct fuse_operations dlorean_oper = {
	.getattr	= dlorean_getattr,
	.statfs		= dlorean_statfs,
	.readlink	= dlorean_readlink,
	.opendir	= dlorean_opendir,
	.readdir	= dlorean_readdir,
	.releasedir	= dlorean_releasedir,
	.mknod		= dlorean_mknod,
	.utime		= dlorean_utime,
	.mkdir		= dlorean_mkdir,
	.rmdir		= dlorean_rmdir,
	.unlink		= dlorean_unlink,
	.symlink	= dlorean_symlink,
	.rename		= dlorean_rename,
	.link		= dlorean_link,
	.chmod		= dlorean_chmod,
	.chown		= dlorean_chown,
	.truncate	= dlorean_truncate,
	.release	= dlorean_release,
	.open		= dlorean_open,
	.read		= dlorean_read,
	.write		= dlorean_write,
};

int main(int argc, char *argv[])
{
	char *path;
	int size_c, n;

	real_path = malloc(strlen(argv[1]) + 1);
	strcpy(real_path, argv[1]);
	mount_point_path = malloc(strlen(argv[2]) + 1);
	strcpy(mount_point_path, argv[2]);
	traveling = NULL;

	size_c = strlen(real_path) + 1 + strlen(versions_place) + 5;
	path = malloc(size_c+1);
	sprintf(path,"%s/%s/test",real_path,versions_place);

	n = mknod (path, S_IFREG, S_IFREG);
	if (n == -1) {
		fprintf(stderr, "\nERROR: you can't write on %s/%s/\n"
			"The File System must use that directory, so fix the problem and re-try.\n\n",real_path,versions_place);
		_exit(-1);
	} 
	n = unlink(path);
	free(path);

	fuse_main((argc-1), (argv+1), &dlorean_oper, NULL);
	return 0;
}

